### PR TITLE
Enable SUNDIALS::KINSOL for LinearAlgebra::distributed::Vector

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: source/sundials/copy.cc [hunk 1]
```diff
@@ -56,6 +56,61 @@ namespace SUNDIALS
...
+    void
+    copy(LinearAlgebra::distributed::Vector<double> &dst, const N_Vector &src)
+    {
+      const IndexSet    is = dst.locally_owned_elements();
+      const std::size_t N  = is.n_elements();
+      AssertDimension(N, N_Vector_length(src));
+      dst.zero_out_ghost_values();
+      for (std::size_t i = 0; i < N; ++i)
+        {
+          dst[is.nth_index_in_set(i)] = NV_Ith_P(src, i);
```
Comment (kronbichler) [inline source/sundials/copy.cc:69]: Why don't you use
```suggestion
          dst.local_element(i) = NV_Ith_P(src, i);
```
It seems unnatural to me to transform to global indices here and then back (on the exact same numbering).

File: source/sundials/copy.cc [hunk 2]
```diff
@@ -56,6 +56,61 @@ namespace SUNDIALS
...
+    copy(LinearAlgebra::distributed::BlockVector<double> &dst,
+         const N_Vector &                                 src)
+    {
+      const IndexSet    is = dst.locally_owned_elements();
+      const std::size_t N  = is.n_elements();
+      AssertDimension(N, N_Vector_length(src));
+      dst.zero_out_ghost_values();
+      for (std::size_t i = 0; i < N; ++i)
+        {
+          dst[is.nth_index_in_set(i)] = NV_Ith_P(src, i);
```
Comment (kronbichler) [inline source/sundials/copy.cc:96]: Here for the block vector it should be fine (no other simple way to do it, as far as I see it).

**Review by kronbichler — CHANGES_REQUESTED** [review]

File: source/sundials/copy.cc (hunk 1 above)
Comment (peterrum) [reply source/sundials/copy.cc]: Changed!

File: source/sundials/copy.cc (hunk 2 above)
Comment (peterrum) [reply source/sundials/copy.cc:104]: Indeed! Changed.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**